File: test/mock-client.ts
import type { GitHubClient, GitHubComment, GitHubIssue, GitHubPullRequest, IssueRef } from "../src/types.js";

export interface MockFixture {
  issues: Record<number, GitHubIssue>;
  comments: Record<number, GitHubComment[]>;
  pulls: Record<number, GitHubPullRequest>;
  files: Record<string, string>;
}

export class MockGitHubClient implements GitHubClient {
  constructor(private readonly fixture: MockFixture) {}

  async getIssue(ref: IssueRef): Promise<GitHubIssue> {
    const issue = this.fixture.issues[ref.number];
    if (!issue) throw new Error(`fixture missing issue #${ref.number}`);
    return issue;
  }

  async getIssueComments(ref: IssueRef): Promise<GitHubComment[]> {
    return this.fixture.comments[ref.number] ?? [];
  }

  async getPullRequest(_owner: string, _repo: string, number: number): Promise<GitHubPullRequest> {
    const pr = this.fixture.pulls[number];
    if (!pr) throw new Error(`fixture missing PR #${number}`);
    return pr;
  }

  async getFileContent(_owner: string, _repo: string, path: string): Promise<string | null> {
    return this.fixture.files[path] ?? null;
  }
}

File: python/tests/mock_client.py
"""Ported from test/mock-client.ts: an in-memory GitHubClient implementation
for unit-testing TrackerVerifier and RegexPatternMatcher without any real
network call."""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List, Optional

from shimguard.types import GitHubComment, GitHubIssue, GitHubPullRequest


@dataclass
class MockFixture:
    issues: Dict[int, GitHubIssue] = field(default_factory=dict)
    comments: Dict[int, List[GitHubComment]] = field(default_factory=dict)
    pulls: Dict[int, GitHubPullRequest] = field(default_factory=dict)
    files: Dict[str, str] = field(default_factory=dict)


class MockGitHubClient:
    def __init__(self, fixture: MockFixture) -> None:
        self.fixture = fixture

    def get_issue(self, owner: str, repo: str, number: int) -> GitHubIssue:
        issue = self.fixture.issues.get(number)
        if issue is None:
            raise AssertionError(f"fixture missing issue #{number}")
        return issue

    def get_issue_comments(self, owner: str, repo: str, number: int) -> List[GitHubComment]:
        return self.fixture.comments.get(number, [])

    def get_pull_request(self, owner: str, repo: str, number: int) -> GitHubPullRequest:
        pr = self.fixture.pulls.get(number)
        if pr is None:
            raise AssertionError(f"fixture missing PR #{number}")
        return pr

    def get_file_content(self, owner: str, repo: str, path: str, ref: str = "HEAD") -> Optional[str]:
        return self.fixture.files.get(path)
